processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferGrouper.java

```diff
@@ -0,0 +1,457 @@
+      final int maxBufferGrouperSize
+  )
+  {
+    this.buffer = buffer;
+    this.keySerde = keySerde;
+    this.keySize = keySerde.keySize();
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+    this.maxBufferGrouperSize = maxBufferGrouperSize;
+
+    int offset = HASH_SIZE + keySize;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+
+    this.bucketSize = offset;
+    this.tableArenaSize = buffer.capacity() / (bucketSize + Ints.BYTES) * bucketSize;
+
+    reset();
+  }
+
+  @Override
+  public boolean aggregate(KeyType key, int keyHash)
+  {
+    final ByteBuffer keyBuffer = keySerde.toByteBuffer(key);
+    if (keyBuffer == null) {
+      return false;
+    }
+
+    Preconditions.checkArgument(
+        keyBuffer.remaining() == keySize,
+        "key size[%s] != keySize[%s]",
+        keyBuffer.remaining(),
+        keySize
+    );
+
+    int bucket = findBucket(
+        tableBuffer,
+        buckets,
+        bucketSize,
+        size < Math.min(maxSize, maxBufferGrouperSize),
+        keyBuffer,
+        keySize,
+        keyHash
+    );
+
+    if (bucket < 0) {
+      if (size < maxBufferGrouperSize) {
+        growIfPossible();
+        bucket = findBucket(tableBuffer, buckets, bucketSize, size < maxSize, keyBuffer, keySize, keyHash);
+      }
+
+      if (bucket < 0) {
+        return false;
+      }
+    }
+
+    final int offset = bucket * bucketSize;
+
+    // Set up key if this is a new bucket.
+    if (!isUsed(bucket)) {
+      tableBuffer.position(offset);
+      tableBuffer.putInt(keyHash | 0x80000000);
+      tableBuffer.put(keyBuffer);
+
+      for (int i = 0; i < aggregators.length; i++) {
+        aggregators[i].init(tableBuffer, offset + aggregatorOffsets[i]);
+      }
+
+      buffer.putInt(tableArenaSize + size * Ints.BYTES, offset);
+      size++;
+    }
+
+    // Aggregate the current row.
+    for (int i = 0; i < aggregators.length; i++) {
+      aggregators[i].aggregate(tableBuffer, offset + aggregatorOffsets[i]);
+    }
+
+    return true;
+  }
+
+  @Override
+  public boolean aggregate(final KeyType key)
+  {
+    return aggregate(key, Groupers.hash(key));
+  }
+
+  @Override
+  public void reset()
+  {
+    size = 0;
+    buckets = Math.min(tableArenaSize / bucketSize, INITIAL_BUCKETS);
+
+    if (buckets < 1) {
+      throw new IAE(
+          "Not enough capacity for even one row! Need[%,d] but have[%,d].",
+          bucketSize + Ints.BYTES,
+          buffer.capacity()
+      );
+    }
+
+    maxSize = Math.max(1, (int) (buckets * MAX_LOAD_FACTOR));
+
+    // Clear used bits of new table
+    for (int i = 0; i < buckets; i++) {
+      buffer.put(i * bucketSize, (byte) 0);
+    }
```

Comment:
I will do the following:

- change default initial buckets value to 1024
- make initial buckets configurable through both runtime properties & query context
- modify the growing to use more of the buffer and also able to reuse some space (start halfway through, then grow upwards, then for the final growth start from 0 so we can re-use all but the last table's space)

I also tried some benchmarks for different initialBuckets values:

```
Benchmark                                   (defaultStrategy)  (initialBuckets)  (numProcessingThreads)  (numSegments)  (rowsPerSegment)  (schemaAndQuery)  Mode  Cnt      Score      Error  Units
GroupByBenchmark.querySingleQueryableIndex                 v2                 2                       4              4            100000           basic.A  avgt   25  45976.360 ± 1179.721  us/op
GroupByBenchmark.querySingleQueryableIndex                 v2                32                       4              4            100000           basic.A  avgt   25  45918.088 ± 1242.089  us/op
GroupByBenchmark.querySingleQueryableIndex                 v2              1024                       4              4            100000           basic.A  avgt   25  46076.179 ± 1372.777  us/op
GroupByBenchmark.querySingleQueryableIndex                 v2              8192                       4              4            100000           basic.A  avgt   25  44461.979 ± 1846.974  us/op
GroupByBenchmark.querySingleQueryableIndex                 v2            131072                       4              4            100000           basic.A  avgt   25  42855.991 ± 1134.716  us/op
GroupByBenchmark.querySingleQueryableIndex                 v2        2147483647                       4              4            100000           basic.A  avgt   25  71947.251 ± 1464.970  us/op
```

performance is best when buckets is slightly larger than the result set (makes total sense). Growing has a cost but it is better than allocating a full sized table.